sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/intervalExpressions.scala

```diff
@@ -295,3 +295,43 @@ case class MultiplyYMInterval(
 
   override def toString: String = s"($left * $right)"
 }
+
+// Multiply a day-time interval by a numeric
+case class MultiplyDTInterval(
+    interval: Expression,
+    num: Expression)
+  extends BinaryExpression with ImplicitCastInputTypes with NullIntolerant with Serializable {
+  override def left: Expression = interval
+  override def right: Expression = num
+
+  override def inputTypes: Seq[AbstractDataType] = Seq(DayTimeIntervalType, NumericType)
+  override def dataType: DataType = DayTimeIntervalType
+
+  @transient
+  private lazy val evalFunc: (Long, Any) => Any = right.dataType match {
+    case _: IntegralType => (micros: Long, num) =>
+      Math.multiplyExact(micros, num.asInstanceOf[Number].longValue())
+    case _: DecimalType => (micros: Long, num) =>
+      val decimalRes = ((new Decimal).set(micros) * num.asInstanceOf[Decimal]).toJavaBigDecimal
+      decimalRes.setScale(0, java.math.RoundingMode.HALF_UP).longValueExact()
+    case _: FractionalType => (micros: Long, num) =>
+      Math.round(micros * num.asInstanceOf[Number].doubleValue())
+  }
+
+  override def nullSafeEval(interval: Any, num: Any): Any = {
+    evalFunc(interval.asInstanceOf[Long], num)
+  }
+
+  override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = right.dataType match {
+    case _: IntegralType =>
+      defineCodeGen(ctx, ev, (m, n) => s"java.lang.Math.multiplyExact($m, $n)")
+    case _: DecimalType =>
+      defineCodeGen(ctx, ev, (m, n) =>
+        s"((new Decimal()).set($m).$$times($n)).toJavaBigDecimal()" +
+        ".setScale(0, java.math.RoundingMode.HALF_UP).longValueExact()")
+    case _: FractionalType =>
+      defineCodeGen(ctx, ev, (m, n) => s"java.lang.Math.round($m * (double)$n)")
```

Comment:
Seems that we should fail here too not just round it to the max long value

Reply:
Does the SQL standard require such behavior? From my point of view, we can map `Double.PositiveInfinity` to `Long.MaxValue` as multiple double values can map to the same long value. We just should document such behavior, and borrow the text from `java.lang.Math.round()`
```
     <li>If the argument is negative infinity or any value less than or
       equal to the value of {@code Long.MIN_VALUE}, the result is
       equal to the value of {@code Long.MIN_VALUE}.
     <li>If the argument is positive infinity or any value greater than or
       equal to the value of {@code Long.MAX_VALUE}, the result is
       equal to the value of {@code Long.MAX_VALUE}.</ul>
``` 
@srielau @cloud-fan WDYT?

Reply:
I don't have a strong opinion here. Shall we turn float/double to Decimal and do the calculation?

Reply:
IMO, the interval value expression's return type is an interval and it defines an interval itself. The expression behavior should respect the interval term rather than the multiplier or divisor. According to the standard, we should define  and raise `interval field overflow` rather than numeric overflows when `the number of significant digits exceeds the implementation-defined maximum number of significant digits`.

Reply:
> Shall we turn float/double to Decimal and do the calculation?

Decimal (and Java BigDecimal) doesn't have representation for `NaN`, `PositiveInfinity` and `NegativeInfinity`, see:
```scala
scala> import org.apache.spark.sql.types.Decimal
import org.apache.spark.sql.types.Decimal

scala> Decimal(Double.NaN)
java.lang.NumberFormatException
  at java.math.BigDecimal.<init>(BigDecimal.java:497)

scala> Decimal(Float.PositiveInfinity)
java.lang.NumberFormatException
  at java.math.BigDecimal.<init>(BigDecimal.java:497)

scala> Decimal(Float.MinValue)
res2: org.apache.spark.sql.types.Decimal = -340282346638528860000000000000000000000
```

Reply:
I have found the [roundToLong()](https://guava.dev/releases/19.0/api/docs/com/google/common/math/DoubleMath.html#roundToLong(double,%20java.math.RoundingMode)) from Guava satisfies our needs completely. I am going to invoke it.